### PR TITLE
fix and improve UI

### PR DESCRIPTION
**Pull request by tomasoignons**
For the moment, 

Added buttons next and previous to the bottom of the page, and added the possibility to scroll up to the page.

Improved the navbar, to be smoother, and added the dark blue when you advance (I think it was made doing a pull request).
image

Explicit what train alone and train collaboratively mean (Added a small description to the buttons, optional)
image

Reworked the model library, added error messages when downloading, and added a "confirm" notification when you want to delete a model
image
image

Reworked the about page
image
 the about page

Reworked a bit the form, by center the toogle of the Secure aggregation
image

The detail is in the issue #628, and in particular

 - I think the Previous and Next buttons should be at the bottom of pages rather than top, to follow the user flow who fills information or reads info from top to bottom. Right now users need to scroll up after every page to get to the next step.

 - After clicking on Getting Started, one can hover on the word DISCOllaborative which displays an explanation. This is useful, but it is not clear this is possible to do. Either write the text directly on the website this is essential or at least underline the word to make it clear and make the info appear faster.

- Clicking on Join button (e.g. for titanic task) should get you to the top of the next page, right now it's already scrolled down

- Nothing prevents users from uploading a whole folder rather the individual images, there should be additional checks on what is uploaded.

- NEXT and PREVIOUS buttons should be at the bottom of the page rather than the top

- For the successive steps, color the horizontal bars in dark blue as the user progresses through the tasks and make the current active step different (e.g. bigger circle) than the previous ones

- What does it mean to “Train alone” or “Train collaboratively”? It is not clear for a new user that doesn't know how disco works. Choosing either of them doesn't seem to make a difference from a user point of view

-  Model items do not always have the same width:

-  When turning off the browser model library, trying to download a trained model doesn't trigger any error/feedback, it should state that a user can't download models if the model library is turned off

-  When the model Library is empty, it should state that users can download trained model or need to train models and link where to do so (currently users can download pretrained models in the "evaluate" page)

-  When deleting a model, either ask users to confirm or create an undo button. It is dangerous to be able to misclick and delete a model inadvertently.

-  Order models by tasks and within tasks from most recent to oldest.

-  Graphic design is not coherent with rest of the website, the blue color is a darker shade and the line thickness is thinner.

-  The Evaluate landing page should explain the difference between Test (Data with labels) and Predict (data without labels)

-  Add a loading indicator when downloading a model and a toaster message for successful download (or error)

-  The Evaluation page at the "Connect your data" and "Evaluate your model" should specify if we are currently Testing or Predicting

-  The toggle button doesn't have the usual graphic design, it is in black with double lined border when clicked on

-  About Us menu should have another icon (currently it looks like a profile icon) and the MLO can have a higher quality logo (or smaller size)

-  Improve the text design.

-  "About Us" title should have a title font and size

- Make icons clickable to point to the respective links

-  "Byzantine-Robust Aggregation" for Federated learning and "Secure Aggregation" for Decentralized learning tick boxes shouldn't be centered

-  “Request help on slack” button doesn't do anything

### CODE REVIEW

**Comment by JulienVig** [comment]
Great work!! Could you copy-paste the items from #628 that you addressed in the PR description?

**Comment by tomasoignons** [comment]
These few commits add for all of these little steps : 

- The website remembers at which step you were but sometimes that should be reset. An example would be clearer: Going to Tasks, Joining the Titanic task and clicking next until the Train your Model step gets you to a particular step. Now if you use the sidebar to go back to the Tasks page, and click again on Joining the Titanic task brings you back to the Train your Model step and not the first step (Task Description). This is great but it should not always behave this way. For example, download two pre-trained models from the Evaluate page, let's say Titanic and MNIST. From the model library click on Titanic to Test it, that leads you to the Connect Your Data step. Now from the homepage click on Evaluate. That directly leads you to Connect Your Data step, implying that you are currently testing Titanic while the expected behavior is to let the user choose which model they want to evaluate (since we clicked on the noon-specific Evaluate button).

-  I think the Previous and Next buttons should be at the bottom of pages rather than top, to follow the user flow who fills information or reads info from top to bottom. Right now users need to scroll up after every page to get to the next step.

-  Order models by tasks and within tasks from most recent to oldest.
-  Add a loading indicator when downloading a model and a toaster message for successful download (or error)
-  "Byzantine-Robust Aggregation" for Federated learning and "Secure Aggregation" for Decentralized learning tick boxes shouldn't be centered
-  Choosing Decentralized of Federated displays different Privacy parameters (Secure aggregation for Decentralized, Byzantine aggregation for Federated). However, switching between the two doesn't always update parameters correctly. For example, choosing Decentralized and ticking Secure Aggregation displays the parameter "Maximum Value of Shares used in Secure Aggregation". Now if we change to Federated, it doesn't remove the "Maximum Value of Shares" parameter. Similarly the federated parameter "Percentile of Trusted Users for Byzantine-Robust Aggregation" isn't removed when changing to Decentralized.

- In the Training Parameters, clicking on Add Element makes the button border bug:
Right after clicking on Add Element:

- Overall, there should be more information accessible on parameters, what they do and what the expected format is (float, percentage, integer, string etc). For example, have a question mark icon next to each parameter to display more information on hover.
-  Improve UI design, smaller red cross, rename “trainingParameter” to "Training Parameter" etc
-  Give more explanations on how the collaboration works, how can other people join the custom task after it's created?

All of these are fixing the issue #628, about UI small changes.

There are some errors in the build, but the local tests pass without problem, and the error is from my code (I think).

**Comment by martinjaggi** [comment]
thanks a lot, very cool!

File: webapp/src/components/containers/ButtonCard.vue
Comment (JulienVig) [inline webapp/src/components/containers/ButtonCard.vue:1]: It seems this file is never used, if this is correct you can delete it!

File: webapp/src/components/training/Trainer.vue [hunk 1]
```diff
@@ -15,6 +15,9 @@
           <div class="grid grid-cols-2 gap-8">
             <CustomButton @click="startTraining(false)">
               train alone
+              <template #description>
+                Train on your own
+              </template>
```
Comment (JulienVig) [inline webapp/src/components/training/Trainer.vue]:
```suggestion
```
With the "Train collaboratively" button description I think "train alone" is explicit enough without an additional description

File: webapp/src/components/training/Trainer.vue [hunk 2]
```diff
@@ -24,6 +27,9 @@
               @click="startTraining(true)"
             >
               train collaboratively
+              <template #description>
+                Share the model's weights with other participants
```
Comment (JulienVig) [inline webapp/src/components/training/Trainer.vue]:
```suggestion
                Exchange model updates with other participants
```
I would use less technical terms for non-technical users

File: webapp/src/components/containers/ButtonsCard.vue [hunk 3]
```diff
@@ -39,7 +43,7 @@ import CustomButton from "@/components/simple/CustomButton.vue";
 const props = withDefaults(
   defineProps<{
     titleAlign?: "left" | "center" | "right";
-    buttons: List<readonly [string, () => void]>;
+    buttons: List<readonly [string, () => void, string?]>;
```
Comment (JulienVig) [inline webapp/src/components/containers/ButtonsCard.vue]:
```suggestion
    buttons: List<readonly [string, () => void, string?]>; // button text, on click function, optional description
```
Comment to clarify the different fields

File: webapp/src/components/pages/TaskList.vue [hunk 4]
```diff
@@ -98,6 +98,9 @@ const trainingStore = useTrainingStore()
 const taskStore = useTasksStore()
 const { tasks } = storeToRefs(taskStore)
 
+// For the moment, I only manage to sort the task by doing an iterator, which is forbidden by EsLint and the version
+// An easy way to fix it is to add a flag to disable the rule in the .eslintrc.js file but I will not do it, because
+// It is a "too big" change to the project I think.
```
Comment (JulienVig) [inline webapp/src/components/pages/TaskList.vue]: This comment shouldn't be included in the code, instead you can leave a PR comment for example

File: webapp/src/components/progress_bars/ProgressIcon.vue [hunk 5]
```diff
@@ -11,7 +11,8 @@
           :class="props.active ? 'bg-slate-500' : 'bg-slate-200'" 
         >
           <div
-            class="py-1"
+            class="py-1 rounded"
+            :class="props.active ? 'bg-disco-blue' : 'bg-slate-200'"
           />
```
Comment (JulienVig) [inline webapp/src/components/progress_bars/ProgressIcon.vue]:
```suggestion
          :class="props.active ? 'bg-disco-blue' : 'bg-slate-200'" 
        >
          <div class="py-1" />
```
This is equivalent and simpler imo

File: webapp/src/components/training/Training.vue [hunk 6]
```diff
@@ -38,6 +38,7 @@
       :task="task"
     />
   </div>
+  <TrainingButtons />
```
Comment (JulienVig) [inline webapp/src/components/training/Training.vue:41]: The TrainingBar should also rely on TrainingButtons (as you already did with TestingButtons)

File: webapp/src/components/progress_bars/TestingBar.vue [hunk 7]
```diff
@@ -26,7 +26,7 @@
         @click="handleRoute(1)"
       >
         <template #text>
-          Connect Your Data
+          Connect Your Data ({{ testing }})
```
Comment (JulienVig) [inline webapp/src/components/progress_bars/TestingBar.vue]: I think it would look nicer if we leave "Connect Your Data" and "Evaluate Your Model" as it was and instead replace "Evaluation" line 4 by "Model Evaluation" for testing and "Model Prediction" for prediction. What do you think?

File: webapp/src/components/progress_bars/TrainingButtons.vue [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+
+import { useTrainingStore } from '@/store/training'
+import CustomButton from '@/components/simple/CustomButton.vue'
+
+const router = useRouter()
+const trainingStore = useTrainingStore()
+const route = useRoute()
+
+
```
Comment (JulienVig) [inline webapp/src/components/progress_bars/TrainingButtons.vue]:
```suggestion
```
No need for empty lines :)

File: webapp/src/components/progress_bars/TrainingButtons.vue [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+const scrollToTop = ():void => {
+  const appElement = document.getElementById('scrollable-div');
+  if (appElement) {
+    appElement.scrollTop = 0;
+  }
+}
+
```
Comment (JulienVig) [inline webapp/src/components/progress_bars/TrainingButtons.vue:43]:
```suggestion
```

File: webapp/src/components/sidebar/ModelLibrary.vue [hunk 10]
```diff
@@ -24,31 +23,32 @@
           </span>
           <div v-if="memoryStore.useIndexedDB" class="space-y-4">
             <button
-              v-for="[id, metadata] in memoryStore.models"
+              v-for="[id, metadata] in memoryStore.models.sort((a : ModelMetadata, b : ModelMetadata) => sortModels(a, b))"
```
Comment (JulienVig) [inline webapp/src/components/sidebar/ModelLibrary.vue:26]: I think the model sort should be performed in a `computed` field to react to changes in the memorystore

File: webapp/src/components/sidebar/ModelLibrary.vue [hunk 11]
```diff
@@ -198,6 +198,24 @@ function openTesting(modelID: string) {
...
+    let time = a.hours.replace("h", ":")
+    const dateA = new Date(`${year}-${month}-${day}T${time}`);
+    [day, month, year] = b.date.split("/");
+    time = b.hours.replace("h", ":")
+    const dateB = new Date(`${year}-${month}-${day}T${time}`);
+    return dateB.getTime() - dateA.getTime() // Sort in ascending order
+}
+
+
+function deleteModelConfirm(modelID: string){
```
Comment (JulienVig) [inline webapp/src/components/sidebar/ModelLibrary.vue:213]: I think it would make more sense to allow users to undo the deletion by clicking on the toaster rather than having to confirm it, something like "Click here to undo the model deletion."

If it's too complicated from an implementation point of view we can leave it as it is. I would increase the time from 5s to 10s and reword it into something like "Click here to confirm the model deletion"

File: webapp/src/components/sidebar/SideBar.vue [hunk 12]
```diff
@@ -182,6 +187,9 @@ function goToNewCustomTask () {
 }
 
 function goToEvaluate () {
+  // Reset the store
+  validationStore.step = 0
+  validationStore.model = undefined;
   router.push({ path: '/evaluate' })
```
Comment (JulienVig) [inline webapp/src/components/sidebar/SideBar.vue:193]: Very nice!

File: webapp/src/components/task_creation_form/containers/FileContainer.vue [hunk 13]
```diff
@@ -2,6 +2,8 @@
   <div>
     <div
       class="h-72 my-3"
+      @dragover.prevent="onDragOver"
```
Comment (JulienVig) [inline webapp/src/components/task_creation_form/containers/FileContainer.vue:5]: onDragOver has been commented so this is what's making the build fail

File: webapp/src/components/task_creation_form/TaskForm.vue [hunk 14]
```diff
@@ -50,10 +51,21 @@
...
                 </label>
+                <div>
+                <span
+                  class="hover:cursor-pointer"
+                  v-tippy="{
+                    content: getContent(field),
+                  }"
+                >
+                  <i class="fa fa-question mr-1" />
+                </span>
```
Comment (JulienVig) [inline webapp/src/components/task_creation_form/TaskForm.vue:62]: I really like the hover explanations!
I don't think it is useful to show the field in some cases like for dropdown menu (select), text or checkbox as it is already clear enough. However it is particularly helpful for numerical training and privacy parameters, especially whether we expect an integer or a fraction

Ideally we would also include a textual description of the parameter (what is it used for, how to find the right value etc.) but this can be left as future work.

File: webapp/src/components/task_creation_form/TaskForm.vue [hunk 15]
```diff
@@ -116,6 +128,17 @@
...
         </IconCard>
       </div>
+      <IconCard>
+        <template #title> How to join after ?</template>
+        <div>
+          <p>
+            After submitting the form, others will be able to join the task by going into the
+            <span @click="goToDiscollaborative()" class="underline font-bold cursor-pointer">DISCOllaborative page</span>, and uploading their private data. You can simulate multiple
+            people by opening a new tab in your browser you can have a detailed explanation About
+            how Disco works <span @click="goToInformation()" class="underline font-bold	cursor-pointer">here</span>.
```
Comment (JulienVig) [inline webapp/src/components/task_creation_form/TaskForm.vue]:
```suggestion
            After submitting the form, others will be able to join the task from the
            <span @click="goToDiscollaborative()" class="underline font-bold cursor-pointer">DISCOllaborative page</span>. You can find more explanations about
            Disco in the <span @click="goToInformation()" class="underline font-bold cursor-pointer">Information page</span>.
```
I would reword this way! Be careful not to use "upload" as the point of Disco is exactly the opposite: train models without uploading/sharing data. You can use "connect" or "plug in" for example

File: webapp/src/components/task_creation_form/containers/FileContainer.vue [hunk 16]
```diff
@@ -93,14 +118,48 @@ const emit = defineEmits<Emits>()
...
   if (files === null) return
 
   // fill in the vee-field to trigger yup validation
   fileName.value = (files.length > 0) ? files[0].name : ''
   emit('input', files)
+  toast.success("Upload complete")
+}
+// function onDragOver(event: DragEvent): void {
+//   // Optional: Add visual feedback
+// }
```
Comment (JulienVig) [inline webapp/src/components/task_creation_form/containers/FileContainer.vue]: Make sure not to merge commented code :)

File: webapp/src/components/task_creation_form/containers/FileContainer.vue [hunk 17]
```diff
@@ -93,14 +118,48 @@ const emit = defineEmits<Emits>()
...
+  } else {
+    toast.error("Upload failed, please try again")
+  }
+}
+
+function showUploadingToast() {
+  toast.info("Uploading your data in the browser (your data stays on your computer)")
+}
+
+// clearInput method
```
Comment (JulienVig) [inline webapp/src/components/task_creation_form/containers/FileContainer.vue]: This comment doesn't provide mumch information you can remove it or add more information

**Review by JulienVig — CHANGES_REQUESTED** [review]
Amazing work thanks a lot for your contribution!!

I'm leaving a lot of comments and reviewing the PR quite thoroughly, as you were telling us you would appreciate some feedback on your work, take away what you find useful!

* I'm very impressed by how quickly you onboarded the codebase! Definitely not easy to pull off
* Overall, it is better to create smaller (number of lines changed) and shorter (time between starting and merging) PRs. This PR was opened 2 months ago, while currently we're aiming to merge PRs within ~two weeks. Otherwise the PR trails off too much compared to concurrent work and makes it tedious to keep up to date, to merge and to review. Concretely you can split big issues into smaller ones (#628 is way too big and this is mb) and only include a subset of the changes in one PR. In terms of code changes, this PR size is good, you can create smaller ones but not much bigger than this.
* We usually wait until all github action tests pass before marking the PR as ready and requesting a review, feel free to ask for help if needed! You can tag me in a PR comment or dm me on slack for example
* For the PR title you can simply use the PR objective, for example: "628 emmanuel fix and improve UI" -> "fix and improve UI", no need to include an issue number or your name.
* I'm "requesting changes" because one test fails (I left a comment in FileContainer about this), other than that most of my comments are superficial

Again, thanks for your great work!

File: webapp/src/components/sidebar/ModelLibrary.vue (hunk 10 above)
Comment (tomasoignons) [reply webapp/src/components/sidebar/ModelLibrary.vue:26]: I think there is no need for that, because he already react at the changes in the memorystore when you delete a model, and the ModelLibrary has to be reload if you add or modify a model.

Comment (JulienVig) [reply webapp/src/components/sidebar/ModelLibrary.vue:26]: I see perfect!

**Comment by JulienVig** [comment]
Don't hesitate to request a review once you're done with the changes!

File: webapp/src/components/task_creation_form/TaskForm.vue [hunk 18]
```diff
@@ -116,6 +128,16 @@
...
         </IconCard>
       </div>
+      <IconCard>
+        <template #title> How to join after ?</template>
+        <div>
+          <p>
+            After submitting the form, others will be able to join the task from the
+            <span @click="goToDiscollaborative()" class="underline font-bold cursor-pointer">DISCOllaborative page</span>. You can find more explanations about
+            Disco in the <span @click="goToInformation()" class="underline font-bold cursor-pointer">Information page</span>.
+          </p>
```
Comment (JulienVig) [inline webapp/src/components/task_creation_form/TaskForm.vue]: We actually have components for DISCOllaboratives and DISCO! We also have been displaying hyperlinks with `text-blue-400` (we should have created a component for this already...)
```suggestion
            <span @click="goToDiscollaborative()" class="underline text-blue-400 cursor-pointer"><DISCOllaboratives /> page</span>. You can find more explanations about
            <DISCO /> in the <span @click="goToInformation()" class="underline text-blue-400 cursor-pointer">Information page</span>.
```
Make sure to import the components:
```
import DISCO from "@/components/simple/DISCO.vue";
import DISCOllaboratives from "@/components/simple/DISCOllaboratives.vue";
```

**Review by JulienVig — APPROVED** [review]

**Review by JulienVig — APPROVED** [review]
Great work!! Just a small design comment but you can merge the PR whenever!

**Comment by martinjaggi** [comment]
super cool, thanks!!
